services/src/handlers/datasets.rs

```diff
@@ -97,8 +139,189 @@ fn adjust_user_path_to_upload_path(meta: &mut MetaDataDefinition, upload: &Uploa
     Ok(())
 }
 
+pub(crate) fn auto_create_dataset_handler<C: Context>(
+    ctx: C,
+) -> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone {
+    warp::path!("dataset" / "auto")
+        .and(warp::post())
+        .and(authenticate(ctx.clone()))
+        .and(warp::any().map(move || ctx.clone()))
+        .and(warp::body::json())
+        .and_then(auto_create_dataset)
+}
+
+// TODO: move into handler once async closures are available?
+async fn auto_create_dataset<C: Context>(
+    session: Session,
+    ctx: C,
+    create: AutoCreateDataSet,
+) -> Result<impl warp::Reply, warp::Rejection> {
+    let upload = ctx
+        .data_set_db_ref()
+        .await
+        .get_upload(session.user.id, create.upload)
+        .await?;
+
+    let create = create.validated()?.user_input;
+
+    let main_file_path = upload.id.root_path()?.join(&create.main_file);
+    let meta_data = auto_detect_dataset(&main_file_path)?;
+
+    let properties = AddDataSet {
+        id: None,
+        name: create.dataset_name,
+        description: create.dataset_description,
+        source_operator: "OgrSource".to_owned(),
+    };
+
+    let mut db = ctx.data_set_db_ref_mut().await;
+    let meta_data = db.wrap_meta_data(meta_data);
+    let id = db
+        .add_data_set(session.user.id, properties.validated()?, meta_data)
+        .await?;
+
+    Ok(warp::reply::json(&IdResponse::from(id)))
+}
+
+fn auto_detect_dataset(main_file_path: &Path) -> Result<MetaDataDefinition> {
+    let mut dataset = Dataset::open(&main_file_path).context(error::Gdal)?;
+    let layer = {
+        if let Ok(layer) = dataset.layer(0) {
+            layer
+        } else {
+            // TODO: handle Raster datasets as well
```

Comment:
Don't raster datasets have layers as well? They have no features, but layers I guess.

Reply:
well the layer() method calls some OGR_ method internally so I figure you only use this for vector datasets and use rasterbands/channels for rasters